### PR TITLE
feat: add eval test cases for empty graph and isImpliedBy

### PR DESCRIPTION
**Pull request by jeswr**
Test cases addressing:
 - https://github.com/w3c/N3/issues/185
 - https://github.com/w3c/N3/issues/157
 
There may be a problem with some of these test cases as some:
 - Contain literals in the subject position
 - Contain variables
 
 Neither of which is actually permitted in the n-quads spec. A solution for those test cases could be to have the same content for the result but contained in another n3 document. What do you think @william-vw ?

### CODE REVIEW

**Comment by william-vw** [comment]
@jeswr 
> A solution for those test cases could be to have the same content for the result but contained in another n3 document.

Just to be sure, you mean keeping the results in an n3 results file instead of n-triples / n-quads?

**Comment by jeswr** [comment]
> Just to be sure, you mean keeping the results in an n3 results file instead of n-triples / n-quads?

Correct

**Comment by william-vw** [comment]
@jeswr yes I think that makes a lot of sense!

File: tests/N3Tests/isImpliedBy/isImpliedBy_bcRule.nq [hunk 1]
```diff
@@ -0,0 +1,6 @@
+?a <http://example.org/b1> <http://example.org/c1> _:b0 .
+?a <http://example.org/b> <http://example.org/c> _:b1 .
+_:b0 <http://www.w3.org/2000/10/swap/log#isImpliedBy> _:b1 _:b2 .
```
Comment (jeswr) [inline tests/N3Tests/isImpliedBy/isImpliedBy_bcRule.nq:3]: > @jeswr yes I think that makes a lot of sense!

So it turns out that doesn't work on files like this because N3 doesn't fully support `n-quads` syntax, in particular having a graph term like this one.

Do you think it is worth opening up discussion as to whether this syntax should be supported in the spec @william-vw @josd .

Comment (josd) [reply tests/N3Tests/isImpliedBy/isImpliedBy_bcRule.nq:3]: Yes @jeswr this makes a lot sense and the latest EYE does an experimental parsing of n-quads.
```
?a <http://example.org/b1> <http://example.org/c1> _:b0 .
?a <http://example.org/b> <http://example.org/c> _:b1 .
_:b0 <http://www.w3.org/2000/10/swap/log#isImpliedBy> _:b1 _:b2 .
```
is now `--pass`'ed as
```
<<_:sk_0 <http://example.org/b1> <http://example.org/c1>>> <http://www.w3.org/2000/10/swap/graph#name> _:e_b0_1.
<<_:sk_1 <http://example.org/b> <http://example.org/c>>> <http://www.w3.org/2000/10/swap/graph#name> _:e_b1_1.
<<_:e_b0_1 <= _:e_b1_1>> <http://www.w3.org/2000/10/swap/graph#name> _:e_b2_1.
```

Comment (jeswr) [reply tests/N3Tests/isImpliedBy/isImpliedBy_bcRule.nq:3]: Generally the results of eval tests are given in an n-quads format (in particular see the [rdf-tests repo](https://github.com/w3c/rdf-tests/tree/main/rdf/rdf11)) as it is straightforward to parse and as such is a good syntax in which to unambiguously provide the baseline result.

In terms of what I'm proposing, it is that the following:

```n3
@prefix : <http://example.org/> .
"a" :b { :c :d :e } .
```

Be considered equivalent to:

```n3
@prefix : <http://example.org/> .
"a" :b _:b0 .
:c :d :e _:b0 .
```

Parsers over in the RDF/JS world like [N3.js](https://github.com/rdfjs/N3.js/) already interpret the first statement as the following RDF/JS quads:


```ts
const result = [
quad(
  literal("a"),
  namedNode("http://example.org/b"),
  blankNode("b0"),
),
quad(
  namedNode("http://example.org/c"),
  namedNode("http://example.org/d"),
  namedNode("http://example.org/e"),
  blankNode("b0"),
)]
```

Comment (josd) [reply tests/N3Tests/isImpliedBy/isImpliedBy_bcRule.nq:3]: > In terms of what I'm proposing, it is that in N3 the following:
> 
> ```
> @prefix : <http://example.org/> .
> "a" :b { :c :d :e } .
> ```
> 
> Be considered equivalent to:
> 
> ```
> @prefix : <http://example.org/> .
> "a" :b _:b0 .
> :c :d :e _:b0 .
> ```

IMHO this is not the same!

Comment (josd) [reply tests/N3Tests/isImpliedBy/isImpliedBy_bcRule.nq:3]: > In terms of what I'm proposing, it is that in N3 the following:
> 
> ```
> @prefix : <http://example.org/> .
> "a" :b { :c :d :e } .
> ```
> 
> Be considered equivalent to:
> 
> ```
> @prefix : <http://example.org/> .
> "a" :b _:b0 .
> :c :d :e _:b0 .
> ```

They are different, the former is a graph term, the latter is a graph by reference.
I have added initial support for n-quads in eye and this example round trips fine:
```
$ cat nq2.n3
@prefix : <http://example.org/> .

"a" :b _:b0 .
:c :d :e _:b0 .
```
```
$ eye --quiet --nope nq2.n3 --pass
@prefix : <http://example.org/>.

"a" :b _:e_b0_1.
:c :d :e _:e_b0_1.
```
One can of course compose graph term with the following rule in
```
$ cat nq3.n3
@prefix log: <http://www.w3.org/2000/10/swap/log#>.
@prefix graph: <http://www.w3.org/2000/10/swap/graph#>.
@prefix : <http://example.org/> .

"a" :b _:b0 .
:c :d :e _:b0 .
:f :g :h _:b0 .

{   ?A ?B ?C.
    ({?D ?E ?F} {?D ?E ?F ?C} ?L) log:collectAllIn ?SCOPE.
    ?G graph:list ?L.
} => {
    ?A ?B ?G.
}.
```
```
$ eye --quiet --nope nq3.n3 --pass
@prefix log: <http://www.w3.org/2000/10/swap/log#>.
@prefix graph: <http://www.w3.org/2000/10/swap/graph#>.
@prefix : <http://example.org/>.

"a" :b _:e_b0_1.
"a" :b {
    :c :d :e.
    :f :g :h.
}.
:c :d :e _:e_b0_1.
:f :g :h _:e_b0_1.
:c :d :e {
    :c :d :e.
    :f :g :h.
}.
:f :g :h {
    :c :d :e.
    :f :g :h.
}.
```

Comment (william-vw) [reply tests/N3Tests/isImpliedBy/isImpliedBy_bcRule.nq:3]: > Generally the results of eval tests are given in an n-quads format (in particular see the [rdf-tests repo](https://github.com/w3c/rdf-tests/tree/main/rdf/rdf11)) as it is straightforward to parse and as such is a good syntax in which to unambiguously provide the expected result of parsing.

Why would n-quads be better to provide expected results - what would the graph label indicate? I'm not finding n-quads result files in the folder you reference (aside from the rdf-n-quads folder, but those are not result files). Not trying to be difficult here, I am just wondering what the added value would be for an individual results file. 

It's a much bigger discussion, for sure, but I feel that graph terms can be used to group triples belonging to the same graph (these could be made "referentially transparent" as well).

**Comment by jeswr** [comment]
I've updated the files @william-vw .

I'll also respond to https://github.com/w3c/N3/pull/202#discussion_r1282267906 in due course when I have time to write a full response.

File: tests/N3Tests/manifest-parser.ttl [hunk 2]
```diff
@@ -1050,15 +1050,15 @@
 :isImpliedBy_graphs
         a              test:TestN3Eval ;
         mf:action      <isImpliedBy/isImpliedBy_graphs.n3> ;
-        mf:result      <isImpliedBy/isImpliedBy_graphs.nq> ;
+        mf:result      <isImpliedBy/isImpliedBy_graphs_result.n3> ;
         mf:name        "isImpliedBy Eval";
         rdfs:comment   "<= should be parsed as log:isImpliedBy in a complex use case";
```
Comment (william-vw) [inline tests/N3Tests/manifest-parser.ttl:1055]: unsure what is meant by complex use case?

**Review by william-vw — APPROVED** [review]
looks good (aside from minor comment)

File: tests/N3Tests/manifest-parser.ttl [hunk 3]
```diff
@@ -1023,6 +1023,46 @@
...
+        mf:name        "isImpliedBy Eval";
+        rdfs:comment   "<= should be parsed as log:isImpliedBy";
+        rdft:approval  rdft:Approved .
+
+:isImpliedBy_graphs
+        a              test:TestN3Eval ;
+        mf:action      <isImpliedBy/isImpliedBy_graphs.n3> ;
+        mf:result      <isImpliedBy/isImpliedBy_graphs_result.n3> ;
+        mf:name        "isImpliedBy Eval";
+        rdfs:comment   "<= should be parsed as log:isImpliedBy in a complex use case";
```
Comment (jeswr) [inline tests/N3Tests/manifest-parser.ttl:1055]:
```suggestion
        rdfs:comment   "<= should be parsed as log:isImpliedBy with graph as premise and object";
```

File: tests/N3Tests/manifest-parser.ttl [hunk 4]
```diff
@@ -1023,6 +1023,46 @@
...
+        mf:name        "isImpliedBy Eval";
+        rdfs:comment   "<= should be parsed as log:isImpliedBy";
+        rdft:approval  rdft:Approved .
+
+:isImpliedBy_graphs
+        a              test:TestN3Eval ;
+        mf:action      <isImpliedBy/isImpliedBy_graphs.n3> ;
+        mf:result      <isImpliedBy/isImpliedBy_graphs_result.n3> ;
+        mf:name        "isImpliedBy Eval";
+        rdfs:comment   "<= should be parsed as log:isImpliedBy with graph as premise and object";
```
Comment (jeswr) [inline tests/N3Tests/manifest-parser.ttl:1055]:
```suggestion
        rdfs:comment   "<= should be parsed as log:isImpliedBy with N3 graphs as subject and object";
```

File: tests/N3Tests/manifest-parser.ttl (hunk 2 above)
Comment (jeswr) [reply tests/N3Tests/manifest-parser.ttl]: Not sure either - resolved in https://github.com/w3c/N3/pull/202/commits/59f2ae74df01baa1ee2c3bad901ce578c644c252.

I don't have permissions to merge on this repo so will leave that to you.

**Review by william-vw — APPROVED** [review]
